falcon/media/json.py

```diff
@@ -92,22 +97,16 @@ async def deserialize_async(self, stream, content_type, content_length):
     def serialize(self, media, content_type):
         result = self.dumps(media)
 
-        try:
-            result = result.encode('utf-8')
-        except AttributeError:  # pragma: nocover
-            # NOTE(kgriffs): Assume that dumps() is non-standard in that it
-            #   returned a byte string.
-            # TODO(kgriffs): This branch is not currently covered since
-            #   orjson testing has been temporarily disabled.
-            pass
+        if self._dumps_str:
+            result = result.encode()
```

Comment:
Actually this is slower than the try-except. if it does not raise the try block has little to no penality:

```
In [1]: class Mmm:
   ...:     def __init__(self):
   ...:         self.option = True
   ...:
   ...:     def base(self):
   ...:         x = 1
   ...:         x += 1
   ...:         return x
   ...:
   ...:     def wtry(self):
   ...:         x = 1
   ...:         try:
   ...:             x += 1
   ...:         except AttributeError:
   ...:             pass
   ...:         return x
   ...:
   ...:     def wif(self):
   ...:         x = 1
   ...:         if self.option:
   ...:             x+=1
   ...:         return x
   ...:

In [2]: ii = Mmm()

In [3]: %timeit ii.base()
145 ns ± 3.24 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: %timeit ii.wtry()
157 ns ± 4.73 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [5]: %timeit ii.wif()
180 ns ± 5.44 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Reply:
this under `Python 3.8.5` on win 10.

@vytas7 can you also reproduce?

Reply:
Good pushback. With this change I was trying to address the issue of someone using a library like orjson for a speed boost, only to be penalized by that try...except. 

The if-statement approach mitigates the problem at the expense of a few extra ns for the common case. Even better might be to do something like this since it would incur no additional overhead as compared to my first attempt at solution:

```py
        result = self.dumps({'message': 'Hello World'})
        if isinstance(result, str):
            self.serialize = self._serialize_s
            self.serialize_async = self._serialize_async_s
        else:
            self.serialize = self._serialize_b
            self.serialize_async = self._serialize_async_b
```

Testing with the Mmm class:

```py
class Mmm:
    def __init__(self):
        self.option = True
        self.base_alt = self.base

    def base(self):
        x = 1
        x += 1
        return x

    def wtry(self):
        x = 1
        try:
            x += 1
        except AttributeError:
            pass
        return x

    def wtry_fail(self):
        x = 1
        try:
            x.z
        except AttributeError:
            pass
        return x        

    def wif(self):
        x = 1
        if self.option:
            x += 1
        return x

    def wii(self):
        x = 1
        if isinstance(x, int):
            x += 1
        return x

m = Mmm()
```

py36

```sh
In [42]: %timeit m.base()
    ...: %timeit m.base_alt()
    ...: %timeit m.wtry()
    ...: %timeit m.wtry_fail()
    ...: %timeit m.wif()
    ...: %timeit m.wii()
146 ns ± 1.53 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
134 ns ± 1.52 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
156 ns ± 1.02 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
593 ns ± 2.01 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
174 ns ± 1.61 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
258 ns ± 4.05 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

py38

```sh
In [15]: %timeit m.base()
    ...: %timeit m.base_alt()
    ...: %timeit m.wtry()
    ...: %timeit m.wtry_fail()
    ...: %timeit m.wif()
    ...: %timeit m.wii()
132 ns ± 1.08 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
132 ns ± 0.442 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
137 ns ± 0.672 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
590 ns ± 4.61 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
161 ns ± 1.53 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
171 ns ± 0.463 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

(Under py36 the `base_alt()` benchmark was consistently faster for some unknown reason.)

Reply:
PyPy3.7 results for good measure:

```sh
In [3]: %timeit m.base()
   ...: %timeit m.base_alt()
   ...: %timeit m.wtry()
   ...: %timeit m.wtry_fail()
   ...: %timeit m.wif()
   ...: %timeit m.wii()
   ...:
1.51 ns ± 0.00286 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.52 ns ± 0.00887 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.54 ns ± 0.0046 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.85 ns ± 0.00496 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.52 ns ± 0.00386 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.56 ns ± 0.0254 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)

In [4]: %timeit m.base()
   ...: %timeit m.base_alt()
   ...: %timeit m.wtry()
   ...: %timeit m.wtry_fail()
   ...: %timeit m.wif()
   ...: %timeit m.wii()
   ...:
1.54 ns ± 0.00436 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.51 ns ± 0.00429 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.52 ns ± 0.00519 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.55 ns ± 0.00607 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.51 ns ± 0.00323 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.51 ns ± 0.00562 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)

In [5]: %timeit m.base()
   ...: %timeit m.base_alt()
   ...: %timeit m.wtry()
   ...: %timeit m.wtry_fail()
   ...: %timeit m.wif()
   ...: %timeit m.wii()
   ...:
1.55 ns ± 0.0113 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.51 ns ± 0.00398 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.54 ns ± 0.0101 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.56 ns ± 0.00592 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.52 ns ± 0.00722 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
1.55 ns ± 0.0127 ns per loop (mean ± std. dev. of 7 runs, 1000000000 loops each)
```

Reply:
PyPy is just "zzzz, i don't care for your benchmarks"

My preference for this would be to just go with the implementation swap you suggested above if it does not have particular drawbacks in cython

Reply:
I haven't looked too deep at this, but at a first glance I would simply pick something along the lines of
```
        result = self.dumps({'message': 'Hello World'})
        if isinstance(result, str):
            self.serialize = self._serialize_s
            self.serialize_async = self._serialize_async_s
        else:
            self.serialize = self._serialize_b
            self.serialize_async = self._serialize_async_b
```

Reply:
OK, I will run with the impl. swap. Turns out it is the fastest under Cython too:

cythonized under py36

```sh
In [11]: %timeit m.base()
    ...: %timeit m.base_alt()
    ...: %timeit m.wtry()
    ...: %timeit m.wtry_fail()
    ...: %timeit m.wif()
    ...: %timeit m.wii()
107 ns ± 0.313 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
95.3 ns ± 0.421 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
114 ns ± 0.729 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
363 ns ± 2.26 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
131 ns ± 0.493 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
107 ns ± 0.452 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

cythonized under py38

```sh
In [4]: %timeit m.base()
   ...: %timeit m.base_alt()
   ...: %timeit m.wtry()
   ...: %timeit m.wtry_fail()
   ...: %timeit m.wif()
   ...: %timeit m.wii()
115 ns ± 0.813 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
98.2 ns ± 1.46 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
128 ns ± 3.23 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
406 ns ± 9.75 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
142 ns ± 2.1 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
117 ns ± 1.56 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```